### PR TITLE
Allows emitting buildInfo when --noEmit is specified

### LINKED ISSUE #38440: Allow --noEmit to override incremental: true from tsconfig.json
## Search Terms

noEmit incremental override type-check

## Suggestion

If `--noEmit` is specified on the command line, ignore the `incremental: true` setting from `tsconfig.json`.

## Use Cases

I want to have a `type-check` npm script that runs `tsc --noEmit`, only for type checking. Whether `incremental` was set to `true` in `tsconfig.json`, shouldn't prevent that, because this isn't a build.

## Examples

Currently, this leads to

> Option 'noEmit' cannot be specified with option 'incremental'

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### PR DESCRIPTION
**Pull request by sheetalkamat**
This change emits buildinfo even if `noEmit` is specified
Now referencing composite project with `noEmit` setting is error instead of erroring on project itself.
This allows to typecheck only run on composite projects just like any other project.
Also as part of this change, there were checker errors that were skipped if `noEmit` is on. Instead now those errors are generated but filtered when someone asks for semanticDiagnostics instead. This allows to preserver the diagnostics across emit and noEmit builds and having to invalidate cache/emit files/make extra pass on the code to get those diagnostics
Fixes #38440

### CODE REVIEW

File: src/compiler/commandLineParser.ts [hunk 1]
```diff
@@ -462,7 +462,6 @@ namespace ts {
         {
             name: "noEmit",
             type: "boolean",
-            affectsEmit: true,
```
Comment (amcasey) [inline src/compiler/commandLineParser.ts:465]: Personally, I find this surprising enough that I'd leave it present but commented out and with an explanation.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -1013,6 +1013,12 @@ namespace ts {
             }
         }
 
+        function errorSkippedOn(key: keyof CompilerOptions, location: Node | undefined, message: DiagnosticMessage, arg0?: string | number, arg1?: string | number, arg2?: string | number, arg3?: string | number): Diagnostic {
```
Comment (amcasey) [inline src/compiler/checker.ts:1016]: I'm having trouble wrapping my head around what this does and would find an explanation helpful.

File: src/compiler/program.ts [hunk 3]
```diff
@@ -3675,6 +3678,11 @@ namespace ts {
         return { diagnostics, sourceMaps: undefined, emittedFiles, emitSkipped: true };
     }
 
+    /*@internal*/
+    export function filterSemanticDiagnotics(diagnostic: readonly Diagnostic[], option: CompilerOptions): readonly Diagnostic[] {
```
Comment (sandersn) [inline src/compiler/program.ts:3682]: should probably be named `diagnostics`

File: src/compiler/checker.ts (hunk 2 above)
Comment (sandersn) [reply src/compiler/checker.ts:1016]: `location, message, arg0, ...` are the existing error logging parameters.

`key: keyof CompilerOptions` is a typesafe way to assert that `key`, while a string, actually refers to a property in `CompilerOptions`. That way the string can safely be used as a property name of `CompilerOptions` in `filterSemanticDiagnostics`:

`filter(diagnostics, d => !d.skippedOn || !options[d.skippedOn])`

This keeps diagnostics that 

1. were not issued by `errorSkippedOn`, so didn't set `skippedOn`
 -OR-
2. do have `skippedOn`, which is guaranteed to be the name of a property in compiler options, say "noEmit". For the "noEmit" example, `!options[d.skippedOn]` is equivalent to `!options.noEmit`.

This meta-programming JS would require an enum plus a switch-case conversion in C# I think. Or the meta-programming facilities there, though the result wouldn't be as typesafe as in Typescript.

File: src/compiler/program.ts (hunk 3 above)
Comment (glen-84) [reply src/compiler/program.ts:3682]: There's also a typo in the function name.

**Pull request merged**